BUG,DOC: `find_peaks_cwt`: reject troughs, improve SNR of actual peaks

antipassive opened the pull request:
* When finding signal level for SNR calculation, use maximum CWT coefficient on the ridge line, not CWT coefficient at the end of the ridge line
* Allow ridge lines with negative signal to be rejected due to low SNR
* Update docs of `min_snr` parameter in `_filter_ridge_lines` and `find_peaks_cwt` accordingly
* Add unit tests for both fixes

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes #23604
Closes #23684
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
This implements the definition of "signal" in signal-to-noise ratio as described in [1],
which is the reference listed in `_peak_finding.py` itself.
As a result, difference in SNR between true positives and false positives increases.
Higher values of the `min_snr` parameter can then be used without loss of true
positives, leading to better precision for `find_peaks_cwt`.
Furthermore, a subset of false positives (appearing at local minima between peaks)
is eliminated, because ridge lines with negative signal can no more meet a positive `min_snr` requirement.

<!--Please explain your changes.-->

#### Additional information
* This change may lead to _worse_ precision for some users before they raise the value of `min_snr` in the call to `find_peaks_cwt`. This is because the SNR of some false positives will also increase a little, possibly exceeding the value of `min_snr` currently in use.
* That the SNR calculation in [1] and SciPy ever differed seems to be an unintended omission: issuing
`git log --follow scipy/signal/_peak_finding.py`
and checking out the earliest listed commits reveals that the odd SNR calculation is already there in commit `3399829`, which dates back to 2011.
<!--Any additional information you think is important.-->

EDIT: add mention of unit tests and the fix to #23684

#### References
[1] Bioinformatics (2006) 22 (17): 2059-2065.
doi:`10.1093/bioinformatics/btl355`

antipassive committed:
BUG,DOC: find_peaks_cwt: fix definition of signal in SNR (#23604)

* When finding signal level for SNR calculation, use maximum CWT
  coefficient found on the ridge line, not CWT coefficient at the end
  of the ridge line
* Update docs of `min_snr` parameter in `_filter_ridge_lines` and
  `find_peaks_cwt` accordingly

antipassive commented:
While running experiments on the "fixed" version of `find_peaks_cwt`, I found another SNR-related problem that I can't afford to leave as it is. I'll open a new issue shortly.

antipassive committed:
BUG: find_peaks_cwt: use non-negative noise measure for SNR (#23684)

* When calculating SNR for ridge lines, take abs() of the initial
  noise vector, not of the final SNR itself. This causes sub-zero
  ridge lines to be filtered out due to their negative SNR.

tylerjereddy commented on scipy/signal/_peak_finding.py:
```diff
@@ -1190,7 +1191,7 @@ def _filter_ridge_lines(cwt, ridge_lines, window_size=None, min_length=None,
         if len(line[0]) < min_length:
             return False
-        snr = abs(cwt[line[0][0], line[1][0]] / noises[line[1][0]])
+        snr = max(cwt[line[0], line[1]]) / noises[line[1][0]]
```
I drafted a regression test and slightly different diff for gh-23684, which I've pasted below the fold. On my branch the test, based on your example in the ticket, fails before and passes after applying the patch. Maybe you could consider adding a similar regression test here since it looks like you're taking care of that one as well?

<details>

```diff
--- a/scipy/signal/_peak_finding.py
+++ b/scipy/signal/_peak_finding.py
@@ -1190,7 +1190,7 @@ def _filter_ridge_lines(cwt, ridge_lines, window_size=None, min_length=None,
     def filt_func(line):
         if len(line[0]) < min_length:
             return False
-        snr = abs(cwt[line[0][0], line[1][0]] / noises[line[1][0]])
+        snr = cwt[line[0][0], line[1][0]] / abs(noises[line[1][0]])
         if snr < min_snr:
             return False
         return True
diff --git a/scipy/signal/tests/test_peak_finding.py b/scipy/signal/tests/test_peak_finding.py
index 8118c505e8..d2b1fed643 100644
--- a/scipy/signal/tests/test_peak_finding.py
+++ b/scipy/signal/tests/test_peak_finding.py
@@ -1,5 +1,4 @@
 import copy
-
 import numpy as np
 import pytest
 from pytest import raises, warns
@@ -911,3 +910,21 @@ class TestFindPeaksCwt:
         found_locs = find_peaks_cwt(test_data, widths)
 
         np.testing.assert_equal(found_locs, 32)
+
+    def test_gh_23684(self):
+        rng = np.random.default_rng(50)
+        x = np.arange(500)
+
+        def lorentz(x, height, width, loc):
+            return height / (1 + np.pow((x-loc)/width, 2))
+
+        # synthesize noisy signal with peaks
+        noise = rng.standard_normal((500,)) * 1e-3
+        signal = (lorentz(x, 0.3, 10, 100) +
+                  lorentz(x, 0.4, 15, 250) +
+                  lorentz(x, 0.2, 5, 400) + noise)
+
+        # find the 3 peaks
+        cwt_scales = np.geomspace(1, 20, 16)
+        peaks = find_peaks_cwt(signal, cwt_scales, min_snr=4, noise_perc=90)
+        xp_assert_equal(peaks, [100, 250, 400])
```

</details>

tylerjereddy replied:
I'm assuming your patch is more sensible than mine, since I'm not a domain expert, so it is mostly the regression test that is worth considering. I do wonder how well this code is tested though, if my poorly-understood patch also works...

antipassive replied:
Yes, I can add a regression test for the signal level.

antipassive replied:
A regression test now exists for both trough exclusion and signal level.

antipassive committed:
BUG: find_peaks_cwt: exclude troughs from detection result

* Replace noise measure patch of previous commit with a better one that
  won't require users to change their min_snr values. This is the 2nd
  and (hopefully) final fix to gh-23684.
* Add unit test to prevent regression

Co-authored-by: Tyler Reddy <tyler.je.reddy@gmail.com>

antipassive committed:
TST: find_peaks_cwt: add test for SNR signal level fix

* Add unit test that checks the signal level of a ridge line.
  This ensures that SNR issue gh-23604 stays fixed.
* Use noise_perc=10 in the trough exclusion test, because this gives
  a more stable noise level than noise_perc=90
* Clarify some comments in the trough exclusion test

antipassive commented:
It turns out my `refguide_check` fails due to an issue that was fixed in #23685. I'll have to rebase and clean up a bit.

antipassive commented:
I cleaned up this work and pushed a new branch. It has the same fixes but no misleading work-in-progress commits. Also, it passes all CI checks, which probably makes it easier to review.

EDIT: Please find the new PR below
